### PR TITLE
fix  emoji in iOS crash

### PR DESCRIPTION
**Pull request by dansunhappy**
## Description

*fix in iOS paste  emoji  crash

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x]  All existing and new tests are passing.
- [x] I The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by dansunhappy** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fflutter%2Fflutter%2Fpull%2F30851) for more info**.

**Comment by dansunhappy** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fflutter%2Fflutter%2Fpull%2F30851) for more info**.

**Comment by dansunhappy** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fflutter%2Fflutter%2Fpull%2F30851) for more info**.

I signed it!

**Comment by dnfield** [comment]
Is there an issue open to discuss this?

This will need tests.

**Comment by dnfield** [comment]
Also, the CLA bot won't recognize your signed CLA because the commits were authored by a different email address than you're signed in with right now.

File: packages/flutter/lib/src/services/text_editing.dart [hunk 1]
```diff
@@ -57,12 +57,25 @@ class TextRange {
   /// The text before this range.
   String textBefore(String text) {
     assert(isNormalized);
+    if(text.isNotEmpty && start > 0 && start < text.length ){
```
Comment (dnfield) [inline packages/flutter/lib/src/services/text_editing.dart]: nit: spacing should be `if (text.isNotEmpty && start > 0 && start < text.length) {`

File: packages/flutter/lib/src/services/text_editing.dart [hunk 2]
```diff
@@ -57,12 +57,25 @@ class TextRange {
...
+      final int startCodeUnit = text.codeUnitAt(start);
```
Comment (dnfield) [inline packages/flutter/lib/src/services/text_editing.dart]: This seems dangerous.  What if the character is more than a single code unit?

File: packages/flutter/lib/src/services/text_editing.dart [hunk 3]
```diff
@@ -57,12 +57,25 @@ class TextRange {
...
+      if( startCodeUnit & 0x8C00 == 0x8C00 ){
```
Comment (dnfield) [inline packages/flutter/lib/src/services/text_editing.dart]: Same spacing nit as above.  This also seems dangerous - Why are we checking for these specific bits to be set on the first code unit?  Are we certain that this impacts all unicode characters with those bits set? Could we somehow document what this magic number means?

**Comment by dansunhappy** [comment]
> Is there an issue open to discuss this?
> 
> This will need tests.
crash： 
1. In TextField Repeatedly paste emoji
2. crash stack
`2019-04-10 16:23:20.697906+0800 Runner[5581:340646] [Bugly]  Trapped uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid JSON message, decoding failed' 
(
	0   CoreFoundation                      0x0000000186210530 <redacted> + 252
	1   libobjc.A.dylib                     0x00000001853eb9f8 objc_exception_throw + 56
	2   CoreFoundation                      0x000000018612a148 <redacted> + 0
	3   Foundation                          0x0000000186bed1c8 <redacted> + 112
	4   Flutter                             0x0000000102f9cb38 -[FlutterJSONMessageCodec decode:] + 472
	5   Flutter                             0x0000000102f9ceb8 -[FlutterJSONMethodCodec decodeMethodCall:] + 68
	6   Flutter                             0x0000000102f9bc90 __45-[FlutterMethodChannel setMethodCallHandler:]_block_invoke + 56
	7   Flutter                             0x0000000102fbae9c _ZNK5shell21PlatformMessageRouter21HandlePlatformMessageEN3fml6RefPtrIN5blink15PlatformMessageEEE + 224
	8   Flutter                             0x0000000102fbefc0 _ZN5shell15PlatformViewIOS21HandlePlatformMessageEN3fml6RefPtrIN5blink15PlatformMessageEEE + 40
	9   Flutter                             0x000000010301783c _ZNSt3__110__function6__funcIZN5shell5Shell29OnEngineHandlePlatformMessageEN3fml6RefPtrIN5blink15PlatformMessageEEEE4$_27NS_9allocatorIS9_EEFvvEEclEv + 80
	10  Flutter                             0x0000000102fcbf24 _ZN3fml15MessageLoopImpl15RunExpiredTasksEv + 756
	11  Flutter                             0x0000000102fcf6c4 _ZN3fml17MessageLoopDarwin11OnTimerFireEP16__CFRunLoopTimerPS0_ + 32
	12  CoreFoundation                      0x00000001861a28f4 <redacted> + 28
	13  CoreFoundation                      0x00000001861a2624 <redacted> + 864
	14  CoreFoundation                      0x00000001861a1e58 <redacted> + 248
	15  CoreFoundation                      0x000000018619cda8 <redacted> + 1844
	16  CoreFoundation                      0x000000018619c354 CFRunLoopRunSpecific + 436
	17  GraphicsServices                    0x000000018839c79c GSEventRunModal + 104
	18  UIKitCore                           0x00000001b260fb68 UIApplicationMain + 212
	19  Runner                              0x000000010088f904 main + 124
	20  libdyld.dylib                       0x0000000185c628e0 <redacted> + 4
)`
Jietu20190410-163733(1)
problem analysis ：
1.Cursor position calculation error while pasting，Insert emoji bytes 
image
@dnfield

**Comment by dnfield** [comment]
I'm not able to reproduce this, and the crash you're seeing seems like it should probably have a fix in the iOS embedding rather than the framework.

Can you provide a sample test (maybe a driver test?) that reproduces this issue?

**Comment by dnfield** [comment]
(I ran a simple app on the iOS simulator with a TextField widget, and repeatedly pasted 𐐷)

**Comment by dansunhappy** [comment]
> (I ran a simple app on the iOS simulator with a TextField widget, and repeatedly pasted 𐐷)

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, v1.4.18, on Mac OS X 10.14.4 18E226, locale zh-Hans-CN)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 10.2)
[✓] Android Studio (version 3.3)
[!] IntelliJ IDEA Community Edition (version 2018.3.3)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Connected device (2 available)
  iOS 12 device，It's not an simulator
test code 

image



IMG_5314
in this case  will crash

**Comment by dnfield** [comment]
What are you doing to get the cursor in the middle of the emoji?

**Comment by dansunhappy** [comment]
> What are you doing to get the cursor in the middle of the emoji?

Random click

File: packages/flutter/lib/src/services/text_editing.dart (hunk 2 above)
Comment (songfei) [reply packages/flutter/lib/src/services/text_editing.dart]: https://en.wikipedia.org/wiki/UTF-16    @dnfield     one character may be two code unit.

**Comment by songfei** [comment]
@dnfield  may be change iphone language to  Chinese Simple  ， try again.

**Comment by songfei** [comment]
`main() {
  String a='𐐷';
  print(a.codeUnits.map((v)=> v.toRadixString(16)));
  String b='宋';
  p`

File: packages/flutter/lib/src/services/text_editing.dart (hunk 3 above)
Comment (songfei) [reply packages/flutter/lib/src/services/text_editing.dart]: this magic number is utf-16  encoding

**Comment by dnfield** [comment]
Some are 4 code points.  I'm trying to find out how to reproduce the crash though.  

/Cc @garyqian @goderbauer

**Comment by dansunhappy** [comment]
> Some are 4 code points. I'm trying to find out how to reproduce the crash though.
> 
> /Cc @GaryQian @goderbauer
### **1.Test code**
```
import 'package:flutter/material.dart';

class MyTextField extends StatelessWidget {
//  MediaQuery
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: TextField(
          maxLines: null,
        ),
      ),
    );
  }
}

void main() async {
  runApp(MaterialApp(
    home: MyTextField(),
  ));
}
```
### **2. environment**
**device**：iOS 12 ，iPhone
**keyboard setting** ： english & emoji

**Pull request closed without merging**